In [1]:
!pip install tensorflow

In [2]:
!pip install keras

In [3]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential,load_model

In [4]:
import pandas as pd
import numpy as np


In [5]:
from google.colab import files
uploaded = files.upload()


Saving forestfires (1).csv to forestfires (1) (1).csv


In [6]:
import io
data = pd.read_csv(io.StringIO(uploaded['forestfires (1).csv'].decode('utf-8')))


In [7]:
data.head()

month  day  FFMC   DMC  ...  monthnov  monthoct  monthsep  size_category
0   mar  fri  86.2  26.2  ...         0         0         0          small
1   oct  tue  90.6  35.4  ...         0         1         0          small
2   oct  sat  90.6  43.7  ...         0         1         0          small
3   mar  fri  91.7  33.3  ...         0         0         0          small
4   mar  sun  89.3  51.3  ...         0         0         0          small

[5 rows x 31 columns]

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

In [9]:
data.isnull().sum()

month            0
day              0
FFMC             0
DMC              0
DC               0
ISI              0
temp             0
RH               0
wind             0
rain             0
area             0
dayfri           0
daymon           0
daysat           0
daysun           0
daythu           0
daytue           0
daywed           0
monthapr         0
monthaug         0
monthdec         0
monthfeb         0
monthjan         0
monthjul         0
monthjun         0
monthmar         0
monthmay         0
monthnov         0
monthoct         0
monthsep         0
size_category    0
dtype: int64

In [10]:
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()
data['month']= label_encoder.fit_transform(data['month']) 
data['day']= label_encoder.fit_transform(data['day']) 
data['size_category']= label_encoder.fit_transform(data['size_category'])
data.head()

month  day  FFMC   DMC  ...  monthnov  monthoct  monthsep  size_category
0      7    0  86.2  26.2  ...         0         0         0              1
1     10    5  90.6  35.4  ...         0         1         0              1
2     10    2  90.6  43.7  ...         0         1         0              1
3      7    0  91.7  33.3  ...         0         0         0              1
4      7    3  89.3  51.3  ...         0         0         0              1

[5 rows x 31 columns]

In [11]:
array = data.values
X = array[:,0:30]
Y = array[:,30]

In [12]:
X

array([[ 7. ,  0. , 86.2, ...,  0. ,  0. ,  0. ],
       [10. ,  5. , 90.6, ...,  0. ,  1. ,  0. ],
       [10. ,  2. , 90.6, ...,  0. ,  1. ,  0. ],
       ...,
       [ 1. ,  3. , 81.6, ...,  0. ,  0. ,  0. ],
       [ 1. ,  2. , 94.4, ...,  0. ,  0. ,  0. ],
       [ 9. ,  5. , 79.5, ...,  1. ,  0. ,  0. ]])

In [13]:
Y

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

In [14]:
# Importing the necessary packages
from sklearn.model_selection import GridSearchCV, KFold
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from tensorflow.keras.optimizers import Adam

In [15]:
from keras.layers import Dropout

In [19]:
def create_model():
    model = Sequential()
    model.add(Dense(8,input_dim = 30,kernel_initializer = 'uniform',activation = 'softmax'))
    model.add(Dropout(0.2))
    model.add(Dense(8,input_dim = 8,kernel_initializer = 'uniform',activation = 'softmax'))
    model.add(Dropout(0.2))
    model.add(Dense(1,activation = 'sigmoid'))
    
    adam = Adam(lr = 0.01)
    model.compile(loss = 'binary_crossentropy',optimizer = adam,metrics = ['acc'])
    return model


In [20]:
# Create the model
model = KerasClassifier(build_fn = create_model,verbose = 0)

In [21]:
batch_size = [10,20,40]
epochs = [10,50,100]

In [22]:
# Make a dictionary of the grid search parameters
param_grid = dict(batch_size = batch_size,epochs = epochs)

In [24]:
from sklearn.preprocessing import StandardScaler
a= StandardScaler()
a.fit(X)
X_standardized=a.transform(X)


In [26]:
# Build and fit the GridSearchCV
grid = GridSearchCV(estimator = model,param_grid = param_grid,cv = KFold(),verbose = 10)
grid_result = grid.fit(X_standardized,Y)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] batch_size=10, epochs=10 ........................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ............ batch_size=10, epochs=10, score=0.971, total=   1.3s
[CV] batch_size=10, epochs=10 ........................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    1.3s remaining:    0.0s


[CV] ............ batch_size=10, epochs=10, score=0.760, total=   1.1s
[CV] batch_size=10, epochs=10 ........................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    2.3s remaining:    0.0s


[CV] ............ batch_size=10, epochs=10, score=0.524, total=   1.1s
[CV] batch_size=10, epochs=10 ........................................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    3.4s remaining:    0.0s


[CV] ............ batch_size=10, epochs=10, score=0.786, total=   1.1s
[CV] batch_size=10, epochs=10 ........................................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    4.5s remaining:    0.0s


[CV] ............ batch_size=10, epochs=10, score=0.738, total=   1.2s
[CV] batch_size=10, epochs=50 ........................................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    5.7s remaining:    0.0s


[CV] ............ batch_size=10, epochs=50, score=0.962, total=   2.5s
[CV] batch_size=10, epochs=50 ........................................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    8.3s remaining:    0.0s


[CV] ............ batch_size=10, epochs=50, score=0.865, total=   2.5s
[CV] batch_size=10, epochs=50 ........................................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:   10.8s remaining:    0.0s


[CV] ............ batch_size=10, epochs=50, score=0.922, total=   2.5s
[CV] batch_size=10, epochs=50 ........................................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:   13.3s remaining:    0.0s


[CV] ............ batch_size=10, epochs=50, score=0.913, total=   2.5s
[CV] batch_size=10, epochs=50 ........................................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:   15.8s remaining:    0.0s


[CV] ............ batch_size=10, epochs=50, score=0.922, total=   2.4s
[CV] batch_size=10, epochs=100 .......................................
[CV] ........... batch_size=10, epochs=100, score=1.000, total=   4.3s
[CV] batch_size=10, epochs=100 .......................................
[CV] ........... batch_size=10, epochs=100, score=0.885, total=   4.2s
[CV] batch_size=10, epochs=100 .......................................
[CV] ........... batch_size=10, epochs=100, score=0.942, total=   4.5s
[CV] batch_size=10, epochs=100 .......................................
[CV] ........... batch_size=10, epochs=100, score=0.913, total=   4.2s
[CV] batch_size=10, epochs=100 .......................................
[CV] ........... batch_size=10, epochs=100, score=0.893, total=   4.3s
[CV] batch_size=20, epochs=10 ........................................
[CV] ............ batch_size=20, epochs=10, score=0.971, total=   0.9s
[CV] batch_size=20, epochs=10 ........................................
[CV] .

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:  1.4min finished


In [27]:
# Summarize the results
print('Best : {}, using {}'.format(grid_result.best_score_,grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
  print('{},{} with: {}'.format(mean, stdev, param))

Best : 0.9264376282691955, using {'batch_size': 10, 'epochs': 100}
0.755862581729889,0.1423064576338023 with: {'batch_size': 10, 'epochs': 10}
0.916840922832489,0.03073656779795145 with: {'batch_size': 10, 'epochs': 50}
0.9264376282691955,0.04169002414230783 with: {'batch_size': 10, 'epochs': 100}
0.734522032737732,0.14396166479160086 with: {'batch_size': 20, 'epochs': 10}
0.899271833896637,0.054673643678455675 with: {'batch_size': 20, 'epochs': 50}
0.9051531076431274,0.06305617829736168 with: {'batch_size': 20, 'epochs': 100}
0.7305825233459473,0.15435061319000673 with: {'batch_size': 40, 'epochs': 10}
0.864432406425476,0.058494830305331515 with: {'batch_size': 40, 'epochs': 50}
0.9148805141448975,0.060902980150428616 with: {'batch_size': 40, 'epochs': 100}


we get accuracy of 93 % approximately for batch_size= 10, epochs= 100, learning rate=0.01, dropout rate=0.2 and activation function of softmax